File: Wicked Little Devil/Objects/Player.m
//
//  Player.m
//  Wicked Little Devil
//
//  Created by Andrew Girvan on 25/05/2012.
//  Copyright 2012 Wicked Little Websites. All rights reserved.
//

#import "Player.h"

@implementation Player

-(id) initWithTexture:(CCTexture2D*)texture rect:(CGRect)rect
{
    if( (self=[super initWithTexture:texture rect:rect]))
    {
        self.velocity = ccp ( 0 , 0 );
        self.drag = 4;
        self.per_collectable = 10;
        self.collectable_multiplier = 1;
        self.scale = 1.25;
        self.jumpspeed = 5.5;
        if ([[UIScreen mainScreen] bounds].size.height == 568)
        {
            self.jumpspeed = 6;
        }
        self.gravity = 0.18;
        self.modifier_gravity = 0;
        self.health = 1.0;
        self.damage = 1.0;
        self.collected = 0;
        self.bigcollected = 0;
        self.score = 0;
        self.time = 0;
        self.jumps = 0;
        self.deaths = 0;
        self.halocollected = 0;
        self.last_platform_touched = NULL;
        
        // redo
        self.controllable = NO;
        self.toggled_platform = NO;
        self.animating = NO;
        self.falling = NO;
    }
    return self;
}

- (void) move
{
    self.velocity = ccp( self.velocity.x, self.velocity.y - (self.gravity + self.modifier_gravity) );
    self.position = ccp((self.position.x), (self.position.y + self.velocity.y));
    
    if ( self.velocity.y < 0 && self.velocity.y > -5)
    {
        self.falling = NO;
        [self animate:2];
    }

    if ( self.velocity.y < -8.5  && !self.falling )
    {
        self.falling = YES;
        [self animate:3];
    }
}

- (void) jump:(float)speed
{
    self.animating = NO;
    [self animate:1];
    
    self.velocity = ccp (self.velocity.x, speed);
}

- (void) animate:(int)animation_id
{
    if ( !self.animating )
    {
        self.animating = YES;
        
        CCCallFunc *stopAnimation = [CCCallFunc actionWithTarget:self selector:@selector(end_animate)];
        switch (animation_id)
        {
            case 1: // jump
                [self runAction:[CCSequence actions:[CCAnimate actionWithAnimation:self.anim_jump], stopAnimation, nil]];
                break;
            case 2:
                [self runAction:[CCSequence actions:[CCAnimate actionWithAnimation:self.anim_fall], nil]];
                break;
            case 3:
                [self runAction:[CCSequence actions:[CCAnimate actionWithAnimation:self.anim_fallfar], nil]];
                break;
            case 4:
               [self runAction:[CCSpawn actions:[CCAnimate actionWithAnimation:self.anim_die], [CCFadeOut actionWithDuration:0.5f], nil]];
                break;
        }
    }
}

- (void) end_animate
{
    self.animating = NO;
}

- (bool) isAlive
{
    return ( self.health > 0.0 );
}

- (bool) isControllable
{
    return self.controllable;
}

- (void) setupAnimations
{
    [[CCSpriteFrameCache sharedSpriteFrameCache] addSpriteFramesWithFile:@"AnimDevil.plist"];
    CCSpriteBatchNode *spriteSheet = [CCSpriteBatchNode batchNodeWithFile:@"AnimDevil.png"];
    [self addChild:spriteSheet];
    
    NSMutableArray *arr_anim_jump = [NSMutableArray array];
    for(int i = 1; i <= 6; ++i) {
        [arr_anim_jump addObject:
         [[CCSpriteFrameCache sharedSpriteFrameCache] spriteFrameByName:
          [NSString stringWithFormat:@"jump%i.png", i]]];
    }
    
    NSMutableArray *arr_fall = [NSMutableArray array];
    for(int i = 1; i <= 6; ++i) {
        [arr_fall addObject:
         [[CCSpriteFrameCache sharedSpriteFrameCache] spriteFrameByName:
          [NSString stringWithFormat:@"fall%i.png", i]]];
    }
    
    NSMutableArray *arr_fall_far = [NSMutableArray array];
    for(int i = 1; i <= 6; ++i) {
        [arr_fall_far addObject:
         [[CCSpriteFrameCache sharedSpriteFrameCache] spriteFrameByName:
          [NSString stringWithFormat:@"fall_far%i.png", i]]];
    }
    
    NSMutableArray *arr_die = [NSMutableArray array];
    for(int i = 1; i <= 6; ++i) {
        [arr_die addObject:
         [[CCSpriteFrameCache sharedSpriteFrameCache] spriteFrameByName:
          [NSString stringWithFormat:@"die%i.png", i]]];
    }
    
    self.anim_jump      = [CCAnimation animationWithSpriteFrames:arr_anim_jump  delay:0.05f];
    self.anim_fall      = [CCAnimation animationWithSpriteFrames:arr_fall       delay:0.05f];
    self.anim_fallfar   = [CCAnimation animationWithSpriteFrames:arr_fall_far   delay:0.05f];
    self.anim_die       = [CCAnimation animationWithSpriteFrames:arr_die        delay:0.01f];
}

- (void) setupAnimationsDetective
{
    [[CCSpriteFrameCache sharedSpriteFrameCache] addSpriteFramesWithFile:@"AnimDetectiveDevil.plist"];
    CCSpriteBatchNode *spriteSheet = [CCSpriteBatchNode batchNodeWithFile:@"AnimDetectiveDevil.png"];
    [self addChild:spriteSheet];
    
    NSMutableArray *arr_anim_jump = [NSMutableArray array];
    for(int i = 1; i <= 6; ++i) {
        [arr_anim_jump addObject:
         [[CCSpriteFrameCache sharedSpriteFrameCache] spriteFrameByName:
          [NSString stringWithFormat:@"jump%i.png", i]]];
    }
    
    NSMutableArray *arr_fall = [NSMutableArray array];
    for(int i = 1; i <= 6; ++i) {
        [arr_fall addObject:
         [[CCSpriteFrameCache sharedSpriteFrameCache] spriteFrameByName:
          [NSString stringWithFormat:@"fall%i.png", i]]];
    }
    
    NSMutableArray *arr_fall_far = [NSMutableArray array];
    for(int i = 1; i <= 6; ++i) {
        [arr_fall_far addObject:
         [[CCSpriteFrameCache sharedSpriteFrameCache] spriteFrameByName:
          [NSString stringWithFormat:@"fall_far%i.png", i]]];
    }
    
    NSMutableArray *arr_die = [NSMutableArray array];
    for(int i = 1; i <= 6; ++i) {
        [arr_die addObject:
         [[CCSpriteFrameCache sharedSpriteFrameCache] spriteFrameByName:
          [NSString stringWithFormat:@"die%i.png", i]]];
    }
    
    self.anim_jump      = [CCAnimation animationWithSpriteFrames:arr_anim_jump  delay:0.05f];
    self.anim_fall      = [CCAnimation animationWithSpriteFrames:arr_fall       delay:0.05f];
    self.anim_fallfar   = [CCAnimation animationWithSpriteFrames:arr_fall_far   delay:0.05f];
    self.anim_die       = [CCAnimation animationWithSpriteFrames:arr_die        delay:0.01f];
}

- (void) setupCharacter:(int)character
{
    switch(character)
    {
        case 0: //detective
            [self setupAnimationsDetective];
        break;
//            
//        case 1: // wicked little
//        {
//            [self setupAnimationsDetective];
//        }
//        break;
//            
//        case 2: // super devil
//        {
//            [self setupAnimationsDetective];
//        }
//        break;
//            
//        case 3: // zombie
//        {
//            [self setupAnimationsDetective];
//        }
//        break;
//            
//        case 4: // ninja
//        {
//            [self setupAnimationsDetective];
//        }
//        break;
//            
//        case 5: // pirate
//        {
//            [self setupAnimationsDetective];
//        }
//        break;
    }
}


- (void) setupPowerup:(int)powerup
{
    switch (powerup)
    {
        default: break;
        case 4: //	Bouncy Devil I
            self.jumpspeed += 0.5;
            break;
        case 5: //	Bouncy Devil II
            self.jumpspeed += 1.0;
            break;
        case 6: // Bouncy devil III
            self.jumpspeed += 1.5;
            break;
        case 7: // Feather Devil I
            self.damage = self.damage/2;
            break;
        case 8: // Feather Devil II
            self.damage = self.damage/3;
            break;
        case 9: // Feather Devil III
            self.damage = self.damage/4;
            break;
        case 10: // Quick Devil I
            self.drag = 4.5;
            break;
        case 11: // Quick Devil II
            self.drag = 4.75;
            break;
        case 12: // Quick Devil III
            self.drag = 5.5;
            break;
        case 13: // Tough Devil I
            self.health += 1;
            break;
        case 14: // Tough Devil II
            self.health += 2;
            break;
        case 15: // Tough Devil III
            self.health += 3;
            break;
        case 16: // Tough Devil IV
            self.health += 1000;
            break;
        case 17: // Winning Devil I
            self.per_collectable = 15;
            break;
        case 18: // Winning Devil II
            self.per_collectable = 20;
            break;
        case 19: // Winning Devil III
            self.per_collectable = 30;
            break;
        case 20: // Rich Devil I
            self.collectable_multiplier = 2;
            break;
        case 21: // Rich Devil II
            self.collectable_multiplier = 3;
            break;
        case 22: // Rich Devil III
            self.collectable_multiplier = 5;
            break;
        case 100: // boundplatform 
            break;
        case 101: // bubble pop
            break;
        case 102: // magenet soul
            break;
        case 103: // magenet soul
            break;
        case 104: // it's a dud!
            break;
        case 105: // detective outfit
            [self setupAnimationsDetective];
            break;
    }
}

@end

File: WickedDevilCore/Sources/WickedDevilCore/Player.swift
import Foundation

/// Which visual state the player is in. The original `Player` ran cocos2d
/// `CCAnimate` actions; the domain layer only tracks *which* animation should be
/// playing so the SpriteKit layer can drive the actual frames.
public enum PlayerAnimation: Int, Codable, Sendable, CaseIterable {
    case none = 0
    /// `animate:1`
    case jump = 1
    /// `animate:2` — short fall
    case fall = 2
    /// `animate:3` — long fall
    case fallFar = 3
    /// `animate:4` — death (played with a fade-out in the original)
    case die = 4
}

/// Playable characters, mirroring `Player setupCharacter:`.
public enum PlayerCharacter: Int, Codable, Sendable, CaseIterable {
    case detective = 0
    case pixel = 1
    case zombie = 2
    case ninja = 3
    case pirate = 4
    case angel = 5

    /// The sprite sheet the rendering layer should load. Kept here (rather than
    /// in the renderer) because the mapping is part of the ported game data.
    public var animationAtlasName: String {
        switch self {
        case .detective: return "AnimDetectiveDevil"
        case .pixel: return "PixelAnimDevil"
        case .zombie: return "ZombieAnimDevil"
        case .ninja: return "NinjaAnimDevil"
        case .pirate: return "PirateAnimDevil"
        case .angel: return "AngelAnimDevil"
        }
    }

    /// The default devil used when the player has not bought a character.
    public static let defaultAtlasName = "AnimDevil"
}

/// Powerup identifiers used by `Player setupPowerup:` and the shop scenes.
/// IDs >= 100 came from `Powerups_special.plist`, the rest from `Powerups.plist`.
public enum Powerup: Int, Codable, Sendable, CaseIterable {
    case none = 0
    case bouncyDevilI = 4
    case bouncyDevilII = 5
    case bouncyDevilIII = 6
    case featherDevilI = 7
    case featherDevilII = 8
    case featherDevilIII = 9
    case quickDevilI = 10
    case quickDevilII = 11
    case quickDevilIII = 12
    case toughDevilI = 13
    case toughDevilII = 14
    case toughDevilIII = 15
    case toughDevilIV = 16
    case winningDevilI = 17
    case winningDevilII = 18
    case winningDevilIII = 19
    case richDevilI = 20
    case richDevilII = 21
    case richDevilIII = 22
    /// Platforms are never culled while this is equipped (`GameLayer.m`).
    case boundPlatform = 100
    case bubblePop = 101
    /// Small collectables within an 80pt radius drift towards the player.
    case magnetSoul = 102
    /// Small collectables within a 140pt radius drift towards the player.
    case magnetSoulPlus = 103
    /// Rocket launchers do not fire while this is equipped.
    case dud = 104
    case detectiveOutfit = 105

    /// Radius used by the two magnet powerups (`Collectable.m radiusCheck`).
    public var magnetRadius: Double? {
        switch self {
        case .magnetSoul: return 80
        case .magnetSoulPlus: return 140
        default: return nil
        }
    }
}

/// The player, ported from `Objects/Player.h` / `Player.m`.
///
/// This is a custom kinematic platformer: position/velocity/gravity are
/// integrated by hand once per frame, so the maths ports over unchanged. All
/// `CCSprite` inheritance has been replaced by plain position/size/scale data.
public final class Player {

    // MARK: - Transform

    public var position: Vec2
    /// Unscaled content size. `Platform.intersectCheck` uses the *unscaled*
    /// content size, while `worldBoundingBox` applies `scale`.
    public var contentSize: SizeF
    /// `Player.m` init: `self.scale = 1.25`.
    public var scale: Double

    // MARK: - Physics

    public var velocity: Vec2
    public var health: Double
    /// Damage dealt to breakable platforms; halved/thirded/quartered by the
    /// Feather Devil powerups.
    public var damage: Double
    public var jumpSpeed: Double
    public var gravity: Double
    /// Extra gravity layered on top of `gravity` (used by level modifiers).
    public var modifierGravity: Double
    /// Maximum horizontal movement per frame when following the touch point.
    public var drag: Double

    // MARK: - Run counters

    public var collected: Int
    public var bigCollected: Int
    public var haloCollected: Int
    public var score: Int
    /// Whole seconds elapsed; incremented once per second by the game clock.
    public var time: Int
    public var jumps: Int
    public var deaths: Int
    /// Score awarded per small collectable (raised by Winning Devil).
    public var perCollectable: Int
    /// Small collectables counted per pickup (raised by Rich Devil).
    public var collectableMultiplier: Int

    // MARK: - State

    /// Identity of the last platform landed on, replacing the original's
    /// `id last_platform_touched`.
    public var lastPlatformTouchedID: String?
    public var controllable: Bool
    /// Flip-flop used by toggle-switch platforms (tags 5 / 51 / 52).
    public var toggledPlatform: Bool
    /// True while an animation "owns" the player; `animate:` is a no-op then.
    public var animating: Bool
    public var falling: Bool
    /// True while a bubble is carrying the player upwards.
    public var floating: Bool
    public private(set) var animation: PlayerAnimation
    public var character: PlayerCharacter?
    /// Sprite sheet the renderer should use, updated by `setupCharacter`.
    public private(set) var animationAtlasName: String
    /// The powerup applied by `setupPowerup(_:)`; the world reads it for the
    /// magnet radius and the platform-binding / dud-rocket rules.
    public private(set) var equippedPowerup: Powerup

    public let device: DeviceProfile

    public init(
        position: Vec2 = .zero,
        contentSize: SizeF = Player.defaultContentSize,
        device: DeviceProfile = .standard
    ) {
        self.position = position
        self.contentSize = contentSize
        self.device = device
        self.velocity = .zero
        self.drag = 4
        self.perCollectable = 10
        self.collectableMultiplier = 1
        self.scale = 1.25
        self.jumpSpeed = device.jumpSpeed
        self.gravity = device.gravity
        self.modifierGravity = 0
        self.health = 1.0
        self.damage = 1.0
        self.collected = 0
        self.bigCollected = 0
        self.score = 0
        self.time = 0
        self.jumps = 0
        self.deaths = 0
        self.haloCollected = 0
        self.lastPlatformTouchedID = nil
        self.controllable = false
        self.toggledPlatform = false
        self.animating = false
        self.falling = false
        self.floating = false
        self.animation = .none
        self.character = nil
        self.animationAtlasName = PlayerCharacter.defaultAtlasName
        self.equippedPowerup = .none
    }

    /// Size of the devil sprite frames in `AnimDevil.plist`.
    public static let defaultContentSize = SizeF(width: 28, height: 35)

    /// Equivalent of cocos2d's `worldBoundingBox`, which applies the node
    /// transform (including `scale`) to the content-size rect.
    public var boundingBox: RectF {
        RectF.centered(
            at: position,
            size: SizeF(width: contentSize.width * scale, height: contentSize.height * scale)
        )
    }

    public var isAlive: Bool { health > 0.0 }

    public var isControllable: Bool { controllable }

    // MARK: - Integration

    /// One frame of vertical integration, ported verbatim from `Player move`.
    ///
    /// Horizontal motion is *not* applied here: the original moved the player
    /// horizontally in `GameScene control_player`, see `applyHorizontalControl`.
    public func move() {
        velocity.y -= (gravity + modifierGravity)
        position.y += velocity.y

        if velocity.y < 0 && velocity.y > -5 {
            falling = false
            animate(.fall)
        }

        if velocity.y < -8.5 && !falling {
            falling = true
            animate(.fallFar)
        }
    }

    /// `Player jump:` — sets the vertical velocity directly and restarts the
    /// jump animation.
    public func jump(_ speed: Double) {
        animating = false
        animate(.jump)
        velocity.y = speed
    }

    /// `GameScene control_player` — moves the player towards the touch X,
    /// clamped to +/- `drag` per frame.
    public func applyHorizontalControl(towardX targetX: Double) {
        guard controllable else { return }
        var diff = targetX - position.x
        if diff > drag { diff = drag }
        if diff < -drag { diff = -drag }
        position.x += diff
    }

    // MARK: - Animation state machine

    /// `Player animate:` — only takes effect when no animation is running.
    public func animate(_ animation: PlayerAnimation) {
        guard !animating else { return }
        animating = true
        self.animation = animation
    }

    /// `Player end_animate` — the renderer calls this when the current
    /// animation finishes. In the original only the jump animation had this
    /// callback attached; fall/die animations stayed latched.
    public func endAnimation() {
        animating = false
    }

    /// Convenience used by enemies/hazards: force-restart the death animation.
    public func playDeathAnimation() {
        animating = false
        animate(.die)
    }

    // MARK: - Setup

    /// `Player setupCharacter:` — in the original this swapped sprite sheets.
    public func setupCharacter(_ character: PlayerCharacter) {
        self.character = character
        self.animationAtlasName = character.animationAtlasName
    }

    /// `Player setupPowerup:` — applies the equipped powerup's stat changes.
    /// Unknown IDs are ignored, matching the original `default: break`.
    public func setupPowerup(_ powerup: Powerup) {
        equippedPowerup = powerup
        switch powerup {
        case .bouncyDevilI:
            jumpSpeed += 0.5
        case .bouncyDevilII:
            jumpSpeed += 1.0
        case .bouncyDevilIII:
            jumpSpeed += 1.5
        case .featherDevilI:
            damage = damage / 2
        case .featherDevilII:
            damage = damage / 3
        case .featherDevilIII:
            damage = damage / 4
        case .quickDevilI:
            drag = 4.5
        case .quickDevilII:
            drag = 4.75
        case .quickDevilIII:
            drag = 5.5
        case .toughDevilI:
            health += 1
        case .toughDevilII:
            health += 2
        case .toughDevilIII:
            health += 3
        case .toughDevilIV:
            health += 1000
        case .winningDevilI:
            perCollectable = 15
        case .winningDevilII:
            perCollectable = 20
        case .winningDevilIII:
            perCollectable = 30
        case .richDevilI:
            collectableMultiplier = 2
        case .richDevilII:
            collectableMultiplier = 3
        case .richDevilIII:
            collectableMultiplier = 5
        case .detectiveOutfit:
            setupCharacter(.detective)
        case .none, .boundPlatform, .bubblePop, .magnetSoul, .magnetSoulPlus, .dud:
            break
        }
    }

    /// Convenience overload for raw powerup IDs coming out of `User`.
    public func setupPowerup(rawValue: Int) {
        guard let powerup = Powerup(rawValue: rawValue) else { return }
        setupPowerup(powerup)
    }

    // MARK: - Damage

    /// Applies one point of damage and latches the death animation when the
    /// player runs out of health, matching the enemy/projectile handlers.
    ///
    /// - Note: `Enemy.m` case 2 fell through into case 22 and `GameLayer.m`
    ///   decremented `health` twice on a rocket hit. Both are unintended
    ///   double-decrements, so this port applies exactly one hit.
    @discardableResult
    public func takeHit(_ amount: Double = 1) -> Bool {
        health -= amount
        if health <= 0 {
            playDeathAnimation()
            return true
        }
        return false
    }
}
